### ISSUE TITLE
[Bug] Changing theme (night mode) breaks TTS

### ISSUE THREAD

**Issue by jparta** [ankidroid/Anki-Android#8084]
###### Reproduction Steps

1. Enable TTS
2. Go to Study
3. Change theme from hamburger menu (switch on/off night mode)
4. Choose 'Replay audio' from dots menu
5. Complete review

###### Expected Result
On 'Replay audio', TTS works as it did when presenting the card initially.
After completing the review of the card, TTS reads the next card.

###### Actual Result
On 'Replay audio', a 'Select language' menu comes up. After selecting a language for the TTS, a toast error comes up saying 'No text-to-speech language available (short_code_of_language)'.
On presentation of a new card, the same series of events happens.

###### Debug info
AnkiDroid Version = 2.14.3

Android Version = 10

ACRA UUID = f939c8ba-2220-489e-abb1-d041f0afd05a

Scheduler = std2

Crash Reports Enabled = true

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

Similar, and possibly related to #8085, but here there is visual feedback in the form of the configuration menu.
Exiting Study fixes TTS, but not if it is still in the background (where tapping the 'back' button brings it up immidiately). So re-initialization seems to fix it. What is peculiar to me is that when you flip the switch on the theme, a toast comes up saying that TTS is being initialized, but it clearly breaks in some way.
If I go to Decks from Study and select the previous deck again to back to Study, the language selected does become the new TTS language, even though the toast indicated the selection was unsuccesful.

**Comment by mikehardy** [comment]
Hi there - for this and #8085 are there any workaround steps available? Are there any pre-steps required (i.e., viewing a TTS card prior to the steps you mention)? I ask because I believe the TTS engine is initialized once, perhaps something about the initialization is lost, specifically when you switch to dark mode I believe the Activity is restarted but not the Application, which may affect how the TTS engine is setup / referenced

**Comment by jparta** [comment]
I rewrote the first repro step to make it clearer. Going to Study (question shown) initializes TTS.

**Comment by JG916** [comment]
The issue is caused by a race condition from the recreation of the AbstractFlashcardViewer activity:

1. Start activity A
2. `ReadText.initializeTts()` is called
3. Toggle theme
4. activity B is started
5. `ReadText.initializeTts()` is called once more
6. Activity A's `onDestroy()` runs, which calls `ReadText.releaseTts()`

Activity B creates a new instance of TextToSpeech through `initializeTts()`, but Activity A incorrectly causes `shutdown()` to be run on that TTS instance because it is a static field.

One solution would be passing in a Context to `releaseTts()`  and only 'releasing' the TTS resources if that Context matches the one set through the initialization (stored in `mReviewer`).
 I.E. if context X is the one who setup the TTS, only context X can release it. Re-initializing with context Y will prevent context X from releasing it.

I haven't thought through every scenario yet, but I'm happy to put up a PR with this if needed.
Otherwise I can take a deeper look and find a better solution.

IMO I'm not a fan of this approach (it's almost like activity to activity communication), but the static fields make it hard to circumvent this.

Thoughts?

**Comment by mikehardy** [comment]
That does seem horrible! Great investigation though. 

I just had a quick look and that is some seriously distorted code, the relationships between all the objects that touch ReadText would be very difficult to pick apart. For an edge case like this which won't have a great deal of value to solve vs other issues, I'd be willing to take a PR that just gated on Context being the one you expect, as you mention, if it works after Activity restarts (like night mode toggles) and doesn't affect the main code path

**Comment by JG916** [comment]
Sure thing. I'll see if I can put up a PR up sometime before Wednesday.

It's good to hear your thoughts about 'time vs value', as well. This is quite a rabbit hole, so fortunately I didn't get sucked in too soon.

**Comment by mikehardy** [comment]
Yeah - it's a tough balance, and at the end of the day, you choose to do what you like, which implies what you enjoy. Some people (myself included sometimes!) just love to leave an area all polished up, so I'll let myself go down a rabbit hole. Other times / other people are just looking for biggest impact possible in shortest time, and that argues for careful selection of area to work on and very efficient work. As long as you enjoy the effort and the change is a positive step (even incrementally so...) I'm happy to merge and move on. Always plenty to do 'round here :-)
